### ISSUE TITLE
"Killed" when trying to initiate graphene-sgx

### ISSUE THREAD

**Issue by carllei1** [oscarlab/graphene#1939]
I am trying to run Node-RED in Graphene-SGX but I am struggling to create a manifest file/or Makefile to make it work. Since Node-RED is built on Node.js and I have to listen to port 1880, I used the nodejs-express-server example provided by graphene to create my Makefile and manifest file. I am able to run "make SGX=1 DEBUG=1" ​but when I run ​./pal_loader SGX nodejs.manifest.sgx settings.js 1880, ​I only get the response "Killed". I am confused what "Killed" means and how to fix this problem? Does it have anything to do with "kill -9 $$SERVER_ID" in the Makefile? I removed it from the Makefile in a previous attempt but I still get the response "Killed".

Manifest file:
```
# Show/hide debug log of Graphene ('inline' or 'none' respectively).
loader.debug_type = $(GRAPHENEDEBUG)

# Read application arguments directly from the command line. Don't use this on production!
loader.insecure__use_cmdline_argv = 1

# Specify paths to search for libraries. The usual LD_LIBRARY_PATH syntax
# applies. Paths must be in-Graphene visible paths, not host-OS paths (i.e.,
# paths must be taken from fs.mount.xxx.path, not fs.mount.xxx.uri).
loader.env.LD_LIBRARY_PATH = /lib:$(ARCH_LIBDIR):/usr/$(ARCH_LIBDIR)

# Mount host-OS directory to required libraries (in 'uri') into in-Graphene
# visible directory /lib (in 'path').
fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:$(GRAPHENEDIR)/Runtime

fs.mount.lib2.type = chroot
fs.mount.lib2.path = $(ARCH_LIBDIR)
fs.mount.lib2.uri = file:$(ARCH_LIBDIR)

fs.mount.lib3.type = chroot
fs.mount.lib3.path = /usr/$(ARCH_LIBDIR)
fs.mount.lib3.uri = file:/usr/$(ARCH_LIBDIR)

# Set enclave size to 2GB. Recall that SGX v1 requires to specify enclave size at
# enclave creation time.
sgx.enclave_size = 2G

# Set maximum number of in-enclave threads to 8. Recall that SGX v1 requires to
# specify the maximum number of simultaneous threads at enclave creation time.
sgx.thread_num = 8

# Specify all libraries used by Node.js and its dependencies (including all libs
# which can be loaded at runtime via dlopen).
sgx.trusted_files.ld = file:$(GRAPHENEDIR)/Runtime/ld-linux-x86-64.so.2
sgx.trusted_files.libc = file:$(GRAPHENEDIR)/Runtime/libc.so.6
sgx.trusted_files.libm = file:$(GRAPHENEDIR)/Runtime/libm.so.6
sgx.trusted_files.libdl = file:$(GRAPHENEDIR)/Runtime/libdl.so.2
sgx.trusted_files.librt = file:$(GRAPHENEDIR)/Runtime/librt.so.1
sgx.trusted_files.libutil = file:$(GRAPHENEDIR)/Runtime/libutil.so.1
sgx.trusted_files.libpthread = file:$(GRAPHENEDIR)/Runtime/libpthread.so.0
$(NODEJS_TRUSTED_LIBS)

# JavaScript (trusted)
sgx.trusted_files.javascript = file:settings.js
sgx.allowed_files.modules = file:node_modules
#sgx.allowed_files.modules = file:lib
#Dmitrii told to add
sys.insecure__allow_eventfd = 1

```

Makefile:
```
.INTERMEDIATE: nodejs-trusted-libs
nodejs-trusted-libs: nodejs-deps
        @for F in `cat nodejs-deps`; do \
                P=`ldd $(NODEJS_DIR)nodejs | grep $$F | awk '{print $$3; exit}'`; \
                N=`echo $$F | tr --delete '.' | tr --delete '-' | tr --delete '+'`; \
                echo -n "sgx.trusted_files.$$N = file:$$P\\\\n"; \
        done > $@

nodejs.manifest: nodejs.manifest.template nodejs-trusted-libs
        @sed -e 's|$$(GRAPHENEDIR)|'"$(GRAPHENEDIR)"'|g' \
                -e 's|$$(GRAPHENEDEBUG)|'"$(GRAPHENEDEBUG)"'|g' \
                -e 's|$$(NODEJS_DIR)|'"$(NODEJS_DIR)"'|g' \
                -e 's|$$(NODEJS_TRUSTED_LIBS)|'"`cat nodejs-trusted-libs`"'|g' \
                -e 's|$$(ARCH_LIBDIR)|'"$(ARCH_LIBDIR)"'|g' \
                $< > $@

# Generate SGX-specific manifest, enclave signature, and token for enclave initialization
nodejs.manifest.sgx: nodejs.manifest settings.js
        $(GRAPHENEDIR)/Pal/src/host/Linux-SGX/signer/pal-sgx-sign \
                -exec nodejs \
                -libpal $(GRAPHENEDIR)/Runtime/libpal-Linux-SGX.so \
                -key $(SGX_SIGNER_KEY) \
                -manifest $< -output $@

nodejs.sig: nodejs.manifest.sgx

nodejs.token: nodejs.sig
        $(GRAPHENEDIR)/Pal/src/host/Linux-SGX/signer/pal-sgx-get-token -output $@ -sig $^

nodejs:
        ln -s $(NODEJS_DIR)nodejs $@

pal_loader:
        ln -s $(GRAPHENEDIR)/Runtime/pal_loader $@

.PHONY: check
check: all
        ./pal_loader nodejs.manifest settings.js 1880 & SERVER_ID=$$!; \
        sleep 3; \
        curl localhost:1880 > OUTPUT; \
        kill -9 $$SERVER_ID;
        @grep -q "Hello World!" OUTPUT && echo "[ Success 1/1 ]"
        @rm OUTPUT

.PHONY: clean
clean:
        $(RM) *.manifest *.manifest.sgx *.token *.sig nodejs pal_loader OUTPUT

.PHONY: distclean
distclean: clean

```

**Comment by dimakuv** [comment]
I don't see anything special in your `manifest.template` file.

Could you send the auto-generated `manifesyt.sgx` file output? Also, please use backticks instead of `'''` to denote file contents.

**Comment by dimakuv** [comment]
> I am able to run "make SGX=1 DEBUG=1" ​but when I run ​./pal_loader SGX nodejs.manifest.sgx settings.js 1880, ​I only get the response "Killed". I am confused what "Killed" means and how to fix this problem?

Could you send the exact output of your terminal when you run this command?

Is there anything suspicious in `dmesg`?

I am also not sure what this `Killed` may mean... Maybe the Out-Of-Memory Killer (OOM killer) in Linux is triggered?

**Comment by carllei1** [comment]
The exact output from the terminal is.
```
sgxsev@rise-sgx-vm:~/.node-red$ ./pal_loader SGX nodejs.manifest.sgx settings.js 1880
Killed
```


I checked dmesg and I noticed:

```
[2256690.557314] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/user.slice,task=pal-Linux-SGX,pid=27080,uid=1002

[2256690.557319] Out of memory: Killed process 27080 (pal-Linux-SGX) total-vm:4082508kB, anon-rss:1572544kB, file-rss:4kB, shmem-rss:0kB, UID:1002 pgtables:3292kB oom_score_adj:0
```

Maybe I haven't allocated enough memory for Node-RED to run?

Auto-generated manifest.sgx file output:

```

# DO NOT MODIFY. THIS FILE WAS AUTO-GENERATED.

# Nodejs manifest file example
#
# This manifest was prepared and tested on Ubuntu 18.04.

loader.argv0_override = nodejs

# LibOS layer library of Graphene. There is currently only one implementation,
# so it is always set to libsysdb.so.
loader.preload = file:../graphene/Runtime/libsysdb.so

# Show/hide debug log of Graphene ('inline' or 'none' respectively).
loader.debug_type = inline

# Read application arguments directly from the command line. Don't use this on production!
loader.insecure__use_cmdline_argv = 1

# Specify paths to search for libraries. The usual LD_LIBRARY_PATH syntax
# applies. Paths must be in-Graphene visible paths, not host-OS paths (i.e.,
# paths must be taken from fs.mount.xxx.path, not fs.mount.xxx.uri).
loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/usr//lib/x86_64-linux-gnu

# Mount host-OS directory to required libraries (in 'uri') into in-Graphene
# visible directory /lib (in 'path').
fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:../graphene/Runtime

fs.mount.lib2.type = chroot
fs.mount.lib2.path = /lib/x86_64-linux-gnu
fs.mount.lib2.uri = file:/lib/x86_64-linux-gnu

fs.mount.lib3.type = chroot
fs.mount.lib3.path = /usr//lib/x86_64-linux-gnu
fs.mount.lib3.uri = file:/usr//lib/x86_64-linux-gnu

# Set enclave size to 2GB. Recall that SGX v1 requires to specify enclave size at
# enclave creation time.
sgx.enclave_size = 2G

# Set maximum number of in-enclave threads to 8. Recall that SGX v1 requires to
# specify the maximum number of simultaneous threads at enclave creation time.
sgx.thread_num = 8

# Specify all libraries used by Node.js and its dependencies (including all libs
# which can be loaded at runtime via dlopen).
sgx.trusted_files.ld = file:../graphene/Runtime/ld-linux-x86-64.so.2
sgx.trusted_files.libc = file:../graphene/Runtime/libc.so.6
sgx.trusted_files.libm = file:../graphene/Runtime/libm.so.6
sgx.trusted_files.libdl = file:../graphene/Runtime/libdl.so.2
sgx.trusted_files.librt = file:../graphene/Runtime/librt.so.1
sgx.trusted_files.libutil = file:../graphene/Runtime/libutil.so.1
sgx.trusted_files.libpthread = file:../graphene/Runtime/libpthread.so.0
sgx.trusted_files.libgcc_sso1 = file:/lib/x86_64-linux-gnu/libgcc_s.so.1
sgx.trusted_files.libstdcso6 = file:/usr/lib/x86_64-linux-gnu/libstdc++.so.6


# JavaScript (trusted)
sgx.trusted_files.javascript = file:settings.js
sgx.allowed_files.modules = file:node_modules
#sgx.allowed_files.modules = file:lib
#Dmitrii told to add
sys.insecure__allow_eventfd = 1

# Generated by Graphene

sgx.enable_stats = 0
sgx.isvprodid = 0
sgx.isvsvn = 0
sgx.static_address = 1
sgx.trusted_checksum.exec = a6a17ad43b1c2f3bb1f0ff27f8896417ac27181566122ecc5a500465ad4e15cb
sgx.trusted_checksum.javascript = a16a4a6a99d1f0e5150b550d2b882d32e37f5d6f6a36671a3bbeade8c11b8b6b
sgx.trusted_checksum.ld = 4484742e96e67a91236839508447ce5f8677b2e57b0787581a4b2f3995e22f1e
sgx.trusted_checksum.libc = 3338ecec505418abd23bf9e466107a6ac78c7ed5d795b0de98dbcc683b207a9f
sgx.trusted_checksum.libdl = a1d16ebae2c8a3d5d610f97a66d05755b30e087b5119c2c76ec7992ab4d38980
sgx.trusted_checksum.libgcc_sso1 = 2f0e486606bd477c6a6ad3588176a16099ed05a8bf47cf3bd27b490bea3d59e9
sgx.trusted_checksum.libm = 3c207ca9112abf323d19cfd5a08e71d29a0f8c0aa3c24a52fe3ba41352f7592e
sgx.trusted_checksum.libpthread = ca94fb23cb36d01dc77dfeaccb60e1370eaa022dd6ba969c854b14a3c164e319
sgx.trusted_checksum.librt = 462d5f91752dee1e9357c492a5c502b4130074d8a9e59abe6dae1f35cee96dce
sgx.trusted_checksum.libstdcso6 = a7ce5e3ab89895f02ad5d55f6b2090c3882ba5f4e12fbd9b96f0ed1712e2a1f6
sgx.trusted_checksum.libutil = f86255c9ac6390333f535ebd2b8d677860287054021be566e9ecab663cdeb3f8
sgx.trusted_checksum.preload0 = 0136554833841124dd1725e4a837050d8ecf97fba9bd8130252b94c6e8238448
                                                                                                                                                               
```

**Comment by dimakuv** [comment]
> [2256690.557319] Out of memory: Killed process 27080 (pal-Linux-SGX) total-vm:4082508kB, ...

Ok, this indicates that you have only 4GB of RAM on your system. Notice that node.js manifest uses `sgx.enclave_size = 2G`, so one node.js process uses 2GB of enclave memory (which is backed by your normal RAM). However, node.js also creates a child process (I don't know the reasons why). This means that at some point in time, your node.js application comprises two processes == two SGX enclaves, each 2GB in size. This totals 4GB, and Linux notices that no more RAM is available and kills the offending (Graphene) process.

It looks like you don't have a swap file or a very small swap file. You can fix this issue by:
1. Moving to a more powerful machine with at least 8GB of RAM.
2. Creating/extending your swap file to be at least 2-4GB in size.

**Comment by dimakuv** [comment]
You may wonder why native node.js app runs fine, but not the SGX one. The reason is that SGX encrypts all enclave pages, so each created enclave page is a "unique" page in RAM. Thus, Linux cannot use any deduplication and must assume that all these enclave pages indeed hold unique data. That's why Linux thinks that all 4GB of RAM are exhausted, even though most of your enclave pages are in reality zero-filled.

**Comment by carllei1** [comment]
Ah ok! That explains why the software gets "killed". I will have to try to get a cloud VM with more RAM. Thank you for your help!

**Comment by carllei1** [comment]
I was able to expand the vm memory to 8GB so I am no longer getting the same error. I am now getting a new error when attempting to run Graphene sgx. I am using a similar manifest file to the one given in the nodejs-express-server example yet it is saying I am missing a fileskey. Do you know what is causing this error?

```
 sudo ./pal_loader SGX nodejs.manifest.sgx settings.js 1880
sgx.protected_fileskey not found in manifest, protected files will not be available
WARNING: Using insecure argv source. Don't use this configuration in production!
WARNING: all RDTSC/RDTSCP instructions are emulated (imprecisely) via gettime() syscall.
```

**Comment by mkow** [comment]
This is not an error, just a warning that protected files feature is not available because you haven't configured it. I think its wording could be less confusing though :)

**Comment by carllei1** [comment]
Thanks! I noticed that the problem seems to be a segfault error 4 ""The cause was a user-mode read resulting in no page being found." when I am trying to run graphene-sgx. Every time I try to run Node-RED with graphene, it terminates after a couple of seconds. When I checked dmesg, I noticed:

´´´
[43191.537328] enclave_a[19564]: segfault at c8 ip 0000563456b631f5 sp 00007ffdd7095cc0 error 4 in enclave_a[563456a94000+f9000]

´´´

Does anyone have any experience with what is causing this type of error when running graphene?
